### ISSUE TITLE
External DNS lookups fail with docker-compose embedded dns server

### ISSUE THREAD

**Issue by davethegr8** [docker/compose#4689]
I'm not sure if this is a bug or user configuration issue, but it seems like it might be a corner case with dns, so submitting here.

I've got a very simple scenario where any external dns lookup with a docker compose project fails, while analogous commands with plain docker work. I suspect something in the embedded dns server has a conflict with my network's settings and what docker uses/expects, but I don't know enough to diagnose or troubleshoot.

I've setup https://gist.github.com/davethegr8/f791ed7fac6f3fdc897b6945b422d36a with the relevant code, but I suspect that this might be host/kernel specific due to it not reproducing on other systems I have. The problematic host is Ubuntu 14.04.1, while it works on an OSX 10.11.6 host.

Here's what happens:

I have a docker-compose file that runs two containers, one which pings google.com and one which cats /etc/resolv.conf. 

```
version: "2"
services:
  ping:
    image: python:3.6-alpine
    command: "ping -c 1 google.com"
  dns:
    image: python:3.6-alpine
    command: "cat /etc/resolv.conf"
```

The output looks something like this:

```
search redacted.com
options ndots:2
nameserver 172.18.0.1
nameserver 172.18.0.2
nameserver 10.50.5.1

Creating pingtest_ping_1
Creating pingtest_dns_1
Attaching to pingtest_ping_1, pingtest_dns_1
dns_1   | search redacted.com
dns_1   | nameserver 127.0.0.11
dns_1   | options ndots:2 ndots:0
pingtest_dns_1 exited with code 0
ping_1  | ping: bad address 'google.com'
pingtest_ping_1 exited with code 1
```

But, if I run basically identical commands with plain docker, they both succeed.

```
$ docker run --rm python:3.6-alpine ping -c 1 google.com

PING google.com (172.217.3.174): 56 data bytes
64 bytes from 172.217.3.174: seq=0 ttl=50 time=3.913 ms

--- google.com ping statistics ---
1 packets transmitted, 1 packets received, 0% packet loss
round-trip min/avg/max = 3.913/3.913/3.913 ms

$ docker run --rm python:3.6-alpine cat /etc/resolv.conf

search redacted.com
options ndots:2
nameserver 172.18.0.1
nameserver 172.18.0.2
nameserver 10.50.5.1
```

I've already tried hardcoding dns settings in my docker-compose file, the end result is the same. Any ideas/suggestions?

Version info:

```
$ sudo docker --version
Docker version 17.03.0-ce, build 60ccb22
$ sudo docker-compose --version
docker-compose version 1.11.2, build dfed245
```

**Comment by shin-** [comment]
Those docker commands are missing a relevant piece - they don't connect the container to a custom network.
What happens if you do something like:

```
docker network create updog
docker run --rm --network updog python:3.6-alpine ping -c 1 google.com
```

?

**Comment by davethegr8** [comment]
The result of that is that the `docker ...` commands fail the dns resolution as well. Which would indicate not a compose issue? I'm not sure what it would indicate though, I don't know much about the bits between the network & host.

```
103c370d79e963d8839b9904a41ba774806289a08718bee101bc9860a2663ec0
ping: bad address 'google.com'
search redacted.com
nameserver 127.0.0.11
options ndots:2 ndots:0
updog
Recreating pingtest_ping_1
Recreating pingtest_dns_1
Attaching to pingtest_dns_1, pingtest_ping_1
dns_1   | search redacted.com
dns_1   | nameserver 127.0.0.11
dns_1   | options ndots:2 ndots:0
pingtest_dns_1 exited with code 0
ping_1  | ping: bad address 'google.com'
pingtest_ping_1 exited with code 1
```

_ 🅰️➕ network name btw. Would connect again._

**Comment by shin-** [comment]
Yes, that would indicate it's not a Compose issue, and is more related to Docker Engine networking. Are you using a custom network driver or anything non-default there? What is the output of `docker network inspect updog`?

**Comment by davethegr8** [comment]
It looks something like this:

```
$ sudo docker network inspect updog
[
    {
        "Name": "updog",
        "Id": "c541ca15323b13bcd46b28091208c5a82c00d3dc814ac78784acd76115da3c89",
        "Created": "2017-03-30T13:33:12.51405402-07:00",
        "Scope": "local",
        "Driver": "bridge",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "default",
            "Options": null,
            "Config": [
                {
                    "Subnet": "172.20.0.0/16",
                    "Gateway": "172.20.0.1"
                }
            ]
        },
        "Internal": false,
        "Attachable": false,
        "Containers": {
            "75f949af3fac2e7ef2580fc43cbda2553d0f0eb2cb884928a3d6f8b3047a9d3e": {
                "Name": "pingtest_dns_1",
                "EndpointID": "e6dac0ca28f9469a8ea626036b230f91b06e1dd9cba4d81c75a65828dc864893",
                "MacAddress": "02:42:ac:14:00:02",
                "IPv4Address": "172.20.0.2/16",
                "IPv6Address": ""
            },
            "afa11c8e436064468dc15454c7a2994dcc676286b8b07b42ed4b14c02a61f13a": {
                "Name": "pingtest_ping_1",
                "EndpointID": "989e6d7b6d0be4bced0f668e8552438e8d2558efd2fcb8155ea974a7af3c46a2",
                "MacAddress": "02:42:ac:14:00:03",
                "IPv4Address": "172.20.0.3/16",
                "IPv6Address": ""
            }
        },
        "Options": {},
        "Labels": {}
    }
]
```

**Comment by shin-** [comment]
Thanks!

There's nothing there that really jumps out to me, but you should create an issue on the [docker/docker](https://github.com/docker/docker/issues) bugtracker with that info - the people there definitely know more than I do :)

**Comment by davethegr8** [comment]
Cheers, thanks for your help!

**Comment by zebpalmer** [comment]
_Found this issue high on google while trying to debug a very similar issue. Know it's closed, but hopefully this will help you/others._ 

**Try using the non-alpine python images.**  

I was having a very similar issue and eventually found out that alpine has had issues with some options in resolv.conf. In my case, search domain wasn't working on a python2.7 alpine image. I know others are still complaining about alpine dns issues and specifically `ndots` as seen in the very old, very long, issue linked below. I swapped to one of the debian based python images, with no other changes, and magically all my dns problems were solved. 

for reference:
https://github.com/gliderlabs/docker-alpine/issues/8
